src/config.c

```diff
@@ -173,54 +163,151 @@ config_read_cmd(const char *ipath, const char *set_command_format)
+  free(path);
+
+  if ( ! config_read_cmd(rha->config_path, set_command_format, rha) )
    {
-    free(path);
+    config_free(rha);
     return NULL;
    }
-  snprintf(set_command, set_command_size, set_command_format, path);
 
-  cf = popen(set_command, "r");
-  if (cf == NULL) 
+  config_dir = (char *)malloc(strlen(rha->config_path)+3);
+  sprintf(config_dir, "%s.d", rha->config_path);
+  file_cnt = scandir(config_dir, &file_list, NULL, alphasort);
+  for (i = 0; i < file_cnt; i++)
    {
-    free(set_command);
-    free(path);
-    return NULL;
+    if (file_list[i]->d_name[0] != '.')
+	 {
+      snprintf(full_file, PATH_MAX, "%s/%s", config_dir, file_list[i]->d_name);
+      config_read_cmd(full_file, set_command_format, rha);
+     }
+    free(file_list[i]);
    }
+  free(file_list);
 
-  free(set_command);
+  pwd = getpwuid(geteuid());
+  if (pwd != NULL && pwd->pw_dir != NULL)
+   {
+    snprintf(full_file, PATH_MAX, "%s/.blah/user.config", pwd->pw_dir);
+    test = fopen(full_file, "r");
+	if (test != NULL)
+     {
+      fclose(test);
+      config_read_cmd(full_file, set_command_format, rha);
+     }
+   }
 
-  rha = (config_handle *)malloc(sizeof(config_handle));
-  if (rha == NULL)
+  if ((bp = config_get("blah_bin_directory", rha)) != NULL)
    {
-    fclose(cf);
-    free(path);
+    rha->bin_path = strdup(bp->value);
+   }
+  else
+   {
+    rha->bin_path = (char *)malloc(strlen(install_location)+5);
+    if (rha->bin_path != NULL) sprintf(rha->bin_path,"%s/bin",install_location);
+   }
+  if (rha->bin_path == NULL)
+   {
+    /* Out of memory */
+    config_free(rha);
     return NULL;
    }
 
-  rha->config_path = path;
-  rha->list = NULL;
-  if (install_location != NULL) rha->install_path = strdup(install_location);
-  if (rha->install_path == NULL)
+  if ((bp = config_get("blah_sbin_directory", rha)) != NULL)
+   {
+    rha->sbin_path = strdup(bp->value);
+   }
+  else
+   {
+    rha->sbin_path = (char *)malloc(strlen(install_location)+6);
+    if (rha->sbin_path != NULL) sprintf(rha->sbin_path,"%s/sbin",install_location);
+   }
+  if (rha->sbin_path == NULL)
    {
     /* Out of memory */
-    fclose(cf);
     config_free(rha);
     return NULL;
    }
-  rha->bin_path = NULL; /* These may be filled out of config file contents. */
-  rha->sbin_path = NULL; 
-  rha->libexec_path = NULL;
+
+  if ((bp = config_get("blah_libexec_directory", rha)) != NULL)
+   {
+    rha->libexec_path = strdup(bp->value);
+   }
+  else
+   {
+    /* If libexec has a 'blahp' subdirectory, then the blahp files should
+     * be in there.
+     */
+    rha->libexec_path = (char *)malloc(strlen(install_location)+15);
+    if (rha->libexec_path != NULL)
+     {
+      sprintf(rha->libexec_path,"%s/libexec/blahp",install_location);
+      test_dir = opendir(rha->libexec_path);
+      if (test_dir == NULL)
+       {
+        sprintf(rha->libexec_path,"%s/libexec",install_location);
+       } else {
+        closedir(test_dir);
+       }
```

Comment:
too bad we have to do the opendir/closedir here just to check if the path is a dir that we can open.

Reply:
We could do a stat() to check that it's an existing directory, with optional access() to check if we have permissions for it. This code followed the example of the surrounding code of fopen()/fclose() of the regular config files.

Reply:
Nbd either way.  It looks like it'll work as-is so don't feel pressured to polish it anymore if you're not feeling it.